File: src/parser.cc
#include <nash/parser.h>
#include <iostream>
#include <nash/debug.h>

std::vector <Statement*> Parser::parse(){
	std::vector<Statement*> statements;
	while (!isAtEnd()){
		try{
			statements.push_back(declaration());
		}
		catch (ParserError&e){
			std::cout << "error: cannot continue\n";
			return statements;
		}
		
	}
	return statements;
}

Statement* Parser::declaration(){
	try{
		if(match(TokenType::VAR)) return varDeclaration();
		if (match(TokenType::FUN)) return functionDeclaration();
		return statement();
	}
	catch (ParserError &e){
		synchronize();
		return NULL;
	}
}

Statement* Parser::varDeclaration(){
	Token* var_name = consume(TokenType::IDENTIFIER, "variable requires a name");
	Expression * value = NULL;
	default_types type;
	consume(TokenType::COLON, "expected ':'");
	if (match(TokenType::INT_TYPE)){
		type = default_types::INT;
	}
	else{
		error(peek(), "expected valid variable type");
	}
	if (match(TokenType::EQUAL)){
		value = expression();
	}

	consume(TokenType::SEMICOLON, "expected ';' after expression");
	
	return new VariableStmt(var_name->lexeme, type, value);
}

Statement * Parser::statement(){
	if (match(TokenType::IF)) return ifStatement();
	if (match(TokenType::RETURN)) return returnStatement();
	return expressionStatement();
}

Statement* Parser::returnStatement(){
	Expression * expr = expression();
	consume(TokenType::SEMICOLON, "expected ';' after expression");
	return new ReturnStmt(expr);
}

Statement * Parser::ifStatement(){
	Statement* ifBody = NULL;
	Statement* elseBody = NULL;
	consume(TokenType::LEFT_PAREN, "expected '(' after if.");
	Expression * condition = expression();
	consume(TokenType::RIGHT_PAREN, "expected ')' after condition");
	consume(TokenType::LEFT_BRACE, "expected '{' after if statement");
	ifBody = block();
	if (match(TokenType::ELSE)){
		consume(TokenType::LEFT_BRACE, "expected '{' after else statement");
		elseBody = block();
	}

	return new IfStmt(condition, ifBody, elseBody);

}

Statement* Parser::block(){
	std::vector<Statement*> statements;

	while (!check(TokenType::RIGHT_BRACE) && !isAtEnd()){
		statements.push_back(declaration());
	}
	consume(TokenType::RIGHT_BRACE, "expected '}' after block");

	return new BlockStmt(statements);
}

Statement * Parser::expressionStatement(){
	Expression *expr = expression();
	consume(TokenType::SEMICOLON, "expected ';' after expression");
	return new ExpressionStmt(expr);
}

Expression* Parser::expression(){
	return assignment();
}

Expression* Parser::_or(){
	Expression* expr = _and();

	while(match(TokenType::OR)){
		Token * _operator = previous();
		Expression* right = _and();
		expr = new LogicalExpr(expr, _operator, right);
	}
	return expr;
}

Expression* Parser::_and(){
	Expression* expr = equality();

	while (match(TokenType::AND)){
		Token* _operator = previous();
		Expression* right = equality();
		expr = new LogicalExpr(expr, _operator, right);
	}
	return expr;
}

Expression* Parser::assignment(){
	Expression * expr = _or();

	if (match(TokenType::EQUAL)){
		Token * equals = previous();
		Expression* value = assignment();
		if (expr->type() == ExpreType::VariableExpr){
			Token* name = ((VariableExpr*)expr)->name;
			return new AssignExpr(name, value);
		}
		error(equals, "invalid assignment target");
	}

	return expr;
}

Expression * Parser::equality(){
	Expression * expr = comparison();

	while (match(TokenType::BANG_EQUAL) || match(TokenType::EQUAL_EQUAL)){
		Token *_operator = previous();

		Expression *right = comparison();
		expr = new BinaryExpr(expr, _operator, right);
	}

	return expr;

}

Expression * Parser::comparison(){
	
	Expression * expr = term();

	while (match(TokenType::GREATER) || match(TokenType::GREATER_EQUAL) || match(TokenType::LESS) || match(TokenType::LESS_EQUAL)){
		Token * _operator = previous();
		Expression * right = term();
		expr = new BinaryExpr(expr, _operator, right);
	}

	return expr;
}

Expression *Parser::term(){
	
	Expression *expr = factor();
	while (match(TokenType::MINUS) || match(TokenType::PLUS)){
		Token * _operator = previous();
		Expression *right = factor();
		expr = new BinaryExpr(expr, _operator, right);
	}
	return expr;
}

Expression *Parser::factor(){
	
	Expression * expr = unary();

	while (match(TokenType::SLASH) || match(TokenType::STAR)){
		Token * _operator = previous();
		Expression *right = unary();
		expr = new BinaryExpr(expr, _operator, right);
	}

	return expr;
}

Expression* Parser::call(){
	
	Expression * expr = primary();

	while (true){
		if (match(TokenType::LEFT_PAREN)){
			expr = finishCallExpr(expr);
		}
		else{
			break;
		}
	}

	return expr;
}

Expression* Parser::finishCallExpr(Expression* callee){
	
	std::vector<Expression*> args;
	if (!check(TokenType::RIGHT_BRACE)){
		while (match(TokenType::COMMA)){
			if (args.size() >= 127){
				error(peek(), "can't have more than 127 arguments");
			}
			args.push_back(expression());
		}
	}
	Token * paren = consume(TokenType::RIGHT_PAREN, "expected ')' after arguments");
	return new CallExpr(callee, paren, args);
}

Statement* Parser::functionDeclaration(){
	
	Token * name = consume(TokenType::IDENTIFIER, "expected function name");
	consume(TokenType::LEFT_PAREN, "expected '(' after function name");
	default_types type = default_types::INT;
	
	std::vector<struct param*> params;
	struct param *c_param = NULL;

	if (!check(TokenType::RIGHT_PAREN)){
		do {
			if (params.size() >= 127){
				error(peek(), "can't have more than 127 parameters");
			}
			c_param  = new param();
			c_param->name = consume(TokenType::IDENTIFIER, "expected parameter name")->lexeme;

			consume(TokenType::COLON ,"expected ':' after parameter name");
			if (match(TokenType::INT_TYPE)){
				c_param->type = default_types::INT;
			}
			else{
				error(peek(), "expected paramter type");
			}

			params.push_back(c_param);
		} while (match(TokenType::COMMA));
	}
	consume(TokenType::RIGHT_PAREN, "expected ')' after parameters");
	consume(TokenType::COLON, "expected ':' after parameters");
	if (match(TokenType::INT_TYPE)){
		type = default_types::INT;
	}

	consume(TokenType::LEFT_BRACE, "expect '{' before function body");

	Statement* body = block();
	return new FunctionStmt(name, params, body, type);
}

Expression *Parser::unary(){
	
	if (match(TokenType::BANG) || match(TokenType::MINUS)){
		Token * _operator = previous();
		Expression *right = unary();
		return new UnaryExpr(_operator, right);
	}
	return call();
}

Expression *Parser::primary(){
	
	if (match(TokenType::FALSE)) return new BoolLiteralExpr(false);
	if (match(TokenType::TRUE)) return new BoolLiteralExpr(true);
	if (match(TokenType::STRING)) return new StringLiteralExpr(previous()->lexeme);
	if (match(TokenType::NUMBER)){

		return new NumberLiteralExpr(stoi(previous()->lexeme));
	} 
	if (match(TokenType::IDENTIFIER)){

		return new VariableExpr(previous());
	} 
	if (match(TokenType::LEFT_PAREN)){
		Expression * expr = expression();
		consume(TokenType::RIGHT_PAREN, "Expect ')' after expression");
		return new GroupingExpr(expr);
	}
	
	return NULL;
}

bool Parser::match(TokenType type){
	//
	if (check(type)){
		advance();
		return true;
	}
	else{
		return false;
	}
}

bool Parser::check(TokenType type){
	//
	if (isAtEnd()) return false;
	return peek()->type == type;
}

Token * Parser::advance(){
	//
	if (!isAtEnd()) current++;
    return previous();
}

bool Parser::isAtEnd(){
	//
	return peek()->type == TokenType::_EOF;
}

Token *Parser::peek(){
	//
	return tokens.at(current);
}

Token* Parser::previous(){
	//
	return tokens.at(current - 1);
}

Token * Parser::consume(TokenType type, std::string message){
	//
	if (check(type)) return advance();
	error(peek(), message);
	return NULL;
}
void Parser::error(Token *tok, std::string message){
	//
	std::cout << "parser::error: " << message << "\n";
	std::cout << "line: "<< tok->line << "; ";
	std::cout << "broken code:\n";
	std::string res;
	std::string error = source.substr(tok->s_col);
	auto index = error.find('\n');
	if (index != std::string::npos)
  		res = error.substr(0, index);
	std::cout << res << "\n";

	throw ParserError();
}

void Parser::synchronize(){
	
	advance();

	while (!isAtEnd()){

		if (previous()->type == TokenType::SEMICOLON) return;

		switch (peek()->type){
			case TokenType::CLASS:
			case TokenType::FUN:
			case TokenType::VAR:
			case TokenType::FOR:
			case TokenType::IF:
			case TokenType::WHILE:
			case TokenType::RETURN:
				return;
		}

		advance();
	}

}

File: src/parser/parser.ts
import { ParserError, ThrowFatalError } from "../errors";
import { Token, TokenType } from "../lexer";
import {
    BinaryExpression,
    GroupingExpression,
    LiteralExpression,
    UnaryExpression,
    Expression,
    VariableExpression,
    AssignmentExpression,
    LogicalExpression,
    CallExpression,
    FunctionDeclarationExpression,
    FunctionDeclarationExpressionArgument,
} from "./ast/expressions";
import {
    Statement,
    ExpressionStatement,
    VariableStatement,
    BlockStatement,
    IfStatement,
    WhileStatement,
    ClassStatement,
    ExternalFunctionDefinitionStatement,
    ImportModuleStatement,
    ReturnStatement,
    DeferStatement,
    DataDefinitionBlock,
    StructStatement,
} from "./ast/statements";
import { Type } from "./types";
import { Ast } from "./ast/ast";
import * as crypto from "node:crypto";
import { Scope } from "../scope";
export class Parser {
    tokens: Array<Token>;
    currentTokenIndex: number = 0;
    statements: Array<Statement> = [];
    currentScope: Scope;
    topScope: Scope;
    constructor(tokens: Array<Token>) {
        this.tokens = tokens;
        this.currentScope = new Scope(null);
        this.topScope = this.currentScope;
    }
    parse(): Ast {
        while (!this.isAtEnd()) {
            this.statements.push(this.declaration());
        }
        return new Ast(this.statements, null, this.topScope);
    }

    declaration(): Statement {
        if (this.matchTokens([TokenType.VARIABLE, TokenType.CONST])) {
            return this.variableDeclaration();
        } else if (this.matchTokens([TokenType.EXTERN])) {
            return this.externDeclaration();
        } else if (this.matchTokens([TokenType.IMPORT])) {
            return this.importModule();
        } else if (this.matchTokens([TokenType.RETURN])) {
            return this.returnStatement();
        } /*else if (this.matchTokens([TokenType.CLASS])) {
            return this.classDecleration();
        }*/ else {
            return this.statement();
        }
    }
    returnStatement(): Statement {
        let returnValue: Expression = this.expression();
        this.consume(
            TokenType.SEMICOLON,
            "Expected ';' after return statement"
        );
        return new ReturnStatement(returnValue);
    }
    importModule(): Statement {
        let startToken: Token = this.previousToken();
        let modulePath: string = "";
        do {
            let modulePathPart = this.consume(TokenType.IDENTIFIER).value;
            modulePath += modulePathPart;
            if (this.matchTokens([TokenType.COLON])) {
                this.consume(
                    TokenType.COLON,
                    "Expected '::' after module part"
                );
            } else {
                break;
            }

            modulePath += "/";
        } while (true);
        this.consume(TokenType.SEMICOLON, "Expected ';' at the end of import");
        return new ImportModuleStatement(startToken, modulePath);
    }
    // @ts-ignore
    externDeclaration(): Statement {
        if (this.matchTokens([TokenType.FUNCTION])) {
            return this.externalFunctionDeclaration();
        } else {
            ThrowFatalError(
                "Uknown external declaration type",
                this.currentToken().line,
                this.currentToken().column
            );
        }
    }
    externalFunctionDeclaration(): Statement {
        let type: Type = new Type("void");
        let functionName: string = this.consume(
            TokenType.IDENTIFIER,
            "Expected function name for external function definition"
        ).value;

        this.consume(TokenType.LPARAM, "Expected '(' after function name");
        let argsTypes: Array<Type> = [];

        while (!this.checkToken(TokenType.RPARAM)) {
            do {
                let argType: Type;
                argType = new Type(
                    this.consume(
                        TokenType.IDENTIFIER,
                        "Expected argument type"
                    ).value
                );

                argsTypes.push(argType);
            } while (this.matchTokens([TokenType.COMMA]));
        }

        this.consume(TokenType.RPARAM);
        if (this.matchTokens([TokenType.COLON])) {
            type.setType(
                this.consume(
                    TokenType.IDENTIFIER,
                    "Expected function return type"
                ).value
            );
        }
        this.consume(
            TokenType.SEMICOLON,
            "Expected ';' after external function definition"
        );
        return new ExternalFunctionDefinitionStatement(
            functionName,
            type,
            argsTypes
        );
    }
    classDecleration(): Statement {
        let name = this.consume(TokenType.IDENTIFIER).value;
        this.consume(TokenType.LCURLYBRACES);
        let block = this.blockStatement();

        return new ClassStatement(name, block);
    }
    variableDeclaration(): Statement {
        const isConstant = this.previousToken().type === TokenType.CONST;
        const name: Token = this.consume(
            TokenType.IDENTIFIER,
            "Expected variable name"
        );
        let variableType: Type | null = null;
        if (this.matchTokens([TokenType.COLON])) {
            variableType = new Type(
                this.consume(
                    TokenType.IDENTIFIER,
                    "Expected variable type"
                ).value
            );
        }

        let initValue = null;

        if (this.matchTokens([TokenType.EQUAL])) {
            initValue = this.expression();
        }
        if (!initValue) {
            this.consume(
                TokenType.SEMICOLON,
                "Expected ';' at end of variable declaration"
            );
        } else if (initValue?.expectSemiColon) {
            this.consume(
                TokenType.SEMICOLON,
                "Expected ';' at end of variable declaration"
            );
        }

        return new VariableStatement(
            name.value,
            variableType,
            initValue,
            isConstant
        );
    }
    statement(): Statement {
        if (this.matchTokens([TokenType.LCURLYBRACES])) {
            return this.blockStatement();
        } else if (this.matchTokens([TokenType.IF])) {
            return this.ifStatement();
        } else if (this.matchTokens([TokenType.WHILE])) {
            return this.whileStatement();
        } else if (this.matchTokens([TokenType.DEFER])) {
            return this.deferStatement();
        } else if (this.matchTokens([TokenType.STRUCT])) {
            return this.structStatement();
        }
        return this.expressionStatement();
    }
    structStatement(): Statement {
        const startingToken: Token = this.previousToken();
        const name = this.consume(TokenType.IDENTIFIER, "Expected struct name");
        this.consume(TokenType.LCURLYBRACES, "Expected '{' after name");
        const block = this.dataDefinitionBlock();
        return new StructStatement(startingToken, name, block);
    }
    dataDefinitionBlock(): DataDefinitionBlock {
        //this.consume(TokenType.LCURLYBRACES, "Expect '{'");

        const startingToken: Token = this.previousToken();
        let statements: Array<Statement> = [];
        while (!this.checkToken(TokenType.RCURLYBRACES) && !this.isAtEnd()) {
            let statement = this.declaration();
            if (statement instanceof VariableStatement) {
            } else if (
                statement instanceof ExpressionStatement &&
                statement.expr instanceof FunctionDeclarationExpression
            ) {
                const anonFunction: string = this.generateAnonymousFunction(
                    statement.expr.startToken,
                    statement.expr.type,
                    statement.expr.args,
                    statement.expr.block
                );
                // @ts-ignore
                statement.expr.anonFunction = anonFunction;
            } else {
                ThrowFatalError(
                    "Must be variable declaration or function declaration",
                    startingToken.line,
                    startingToken.column
                );
            }
            statements.push(statement);
        }

        this.consume(
            TokenType.RCURLYBRACES,
            `Expected '}' but got '${JSON.stringify(
                this.currentToken()
            )}' instead.`
        );

        return new DataDefinitionBlock(startingToken, statements);
    }
    deferStatement(): Statement {
        const startingToken: Token = this.previousToken();
        const block: Statement = this.statement();
        return new DeferStatement(startingToken, block);
    }
    whileStatement(): Statement {
        this.consume(TokenType.LPARAM);
        const expr: Expression = this.expression();
        this.consume(TokenType.RPARAM);
        const thenBlock: Statement = this.statement();
        return new WhileStatement(expr, thenBlock);
    }
    ifStatement(): Statement {
        this.consume(TokenType.LPARAM);
        const expr: Expression = this.expression();
        this.consume(TokenType.RPARAM);
        const thenBlock: Statement = this.statement();
        let elseBlock = null;
        if (this.matchTokens([TokenType.ELSE])) {
            elseBlock = this.statement();
        }
        return new IfStatement(expr, thenBlock, elseBlock);
    }
    expressionStatement(): Statement {
        const expr: Expression = this.expression();
        if (expr.expectSemiColon) {
            this.consume(TokenType.SEMICOLON, "Expected ';' after expression");
        }
        return new ExpressionStatement(expr);
    }
    blockStatement(): Statement {
        let statements: Array<Statement> = [];
        while (!this.checkToken(TokenType.RCURLYBRACES) && !this.isAtEnd()) {
            statements.push(this.declaration());
        }

        this.consume(
            TokenType.RCURLYBRACES,
            `Expected '}' but got '${JSON.stringify(
                this.currentToken()
            )}' instead.`
        );

        return new BlockStatement(statements);
    }

    expression(): Expression {
        return this.assignment();
    }
    assignment(): Expression {
        let startToken = this.currentToken();
        const expr: Expression = this.or();
        if (this.matchTokens([TokenType.EQUAL])) {
            const value: Expression = this.equality();
            if (expr instanceof VariableExpression) {
                const name = (expr as VariableExpression).name;
                return new AssignmentExpression(startToken, name, value);
            } else {
                throw new ParserError("Invalid assignment target");
            }
        }
        return expr;
    }
    or(): Expression {
        let startToken = this.currentToken();

        let expr: Expression = this.and();
        if (this.matchTokens([TokenType.OR])) {
            const operator: Token = this.previousToken();
            const right: Expression = this.and();
            expr = new LogicalExpression(startToken, expr, operator, right);
        }
        return expr;
    }
    and(): Expression {
        let startToken = this.currentToken();

        let expr: Expression = this.equality();
        if (this.matchTokens([TokenType.AND])) {
            const operator: Token = this.previousToken();
            const right: Expression = this.equality();
            expr = new LogicalExpression(startToken, expr, operator, right);
        }
        return expr;
    }
    equality(): Expression {
        let startToken = this.currentToken();

        let expr: Expression = this.comparison();
        while (this.matchTokens([TokenType.EQUALEQUAL])) {
            const token: Token = this.previousToken();
            const right: Expression = this.comparison();
            expr = new BinaryExpression(startToken, expr, token, right);
        }
        return expr;
    }
    comparison(): Expression {
        let startToken = this.currentToken();

        let expr: Expression = this.term();
        while (
            this.matchTokens([
                TokenType.GREATER_OR_EQUAL,
                TokenType.LESS_OR_EQUAL,
                TokenType.GREATER,
                TokenType.LESS,
            ])
        ) {
            const operator: Token = this.previousToken();
            const right: Expression = this.term();
            expr = new BinaryExpression(startToken, expr, operator, right);
        }
        return expr;
    }
    term(): Expression {
        let startToken = this.currentToken();

        let expr: Expression = this.factor();

        while (this.matchTokens([TokenType.PLUS, TokenType.MINUS])) {
            const operator: Token = this.previousToken();
            const right: Expression = this.factor();
            expr = new BinaryExpression(startToken, expr, operator, right);
        }
        return expr;
    }
    factor(): Expression {
        let startToken = this.currentToken();

        let expr: Expression = this.unary();

        while (this.matchTokens([TokenType.SLASH, TokenType.STAR])) {
            const operator: Token = this.previousToken();
            const right: Expression = this.unary();
            expr = new BinaryExpression(startToken, expr, operator, right);
        }
        return expr;
    }
    unary(): Expression {
        let startToken = this.currentToken();

        if (this.matchTokens([TokenType.BANG, TokenType.MINUS])) {
            const operator: Token = this.previousToken();
            const right: Expression = this.unary();
            return new UnaryExpression(startToken, operator, right);
        }
        return this.call();
    }
    call(): Expression {
        let startToken = this.currentToken();

        let expr: Expression = this.primary();
        while (true) {
            if (this.matchTokens([TokenType.LPARAM])) {
                let args: Array<Expression> = [];
                if (!this.checkToken(TokenType.RPARAM)) {
                    let argument: Expression;
                    do {
                        if (args.length >= 255) {
                            throw new ParserError(
                                "You cannot pass more than 255 arguments to a function"
                            );
                        }
                        argument = this.expression();
                        args.push(argument);
                    } while (this.matchTokens([TokenType.COMMA]));
                }
                this.consume(TokenType.RPARAM);
                expr = new CallExpression(startToken, expr, args);
            } else {
                break;
            }
        }

        return expr;
    }
    // @ts-ignore
    primary(): Expression {
        let startToken = this.currentToken();

        if (this.matchTokens([TokenType.FALSE])) {
            return new LiteralExpression(startToken, false, "boolean");
        } else if (this.matchTokens([TokenType.TRUE])) {
            return new LiteralExpression(startToken, true, "boolean");
        } else if (this.matchTokens([TokenType.NULL])) {
            return new LiteralExpression(startToken, 0, "integer");
        } else if (this.matchTokens([TokenType.INTEGER])) {
            return new LiteralExpression(
                startToken,
                this.previousToken().value,
                "integer"
            );
        } else if (this.matchTokens([TokenType.STRING])) {
            return new LiteralExpression(
                startToken,
                this.previousToken().value,
                "string"
            );
        } else if (this.matchTokens([TokenType.LPARAM])) {
            const expr: Expression = this.expression();
            this.consume(TokenType.RPARAM);
            return new GroupingExpression(startToken, expr);
        } else if (this.matchTokens([TokenType.IDENTIFIER])) {
            return new VariableExpression(startToken, this.previousToken());
        } else if (this.matchTokens([TokenType.FUNCTION])) {
            return this.functionDeclaration();
        }

        ThrowFatalError(
            `Could not parse expression`,
            this.currentToken().line,
            this.currentToken().column
        );
    }
    generateAnonymousFunction(
        startToken: Token,
        type: Type,
        args: Array<FunctionDeclarationExpressionArgument>,
        block: Statement
    ): string {
        let name = "anonymousFunction" + crypto.randomBytes(20).toString("hex");
        this.statements.push(
            new ExpressionStatement(
                new FunctionDeclarationExpression(
                    startToken,
                    name,
                    args,
                    type,
                    block
                )
            )
        );
        return name;
    }
    functionDeclaration(): Expression {
        let startToken = this.previousToken();
        let type: Type = new Type("function<void>");
        let name: string | null = null;
        let block: Statement;
        if (this.matchTokens([TokenType.LPARAM])) {
        } else if (this.matchTokens([TokenType.IDENTIFIER])) {
            name = this.previousToken().value;
            this.consume(TokenType.LPARAM);
        } else {
            throw new ParserError("Expected '(' or function name");
        }
        let args: Array<FunctionDeclarationExpressionArgument> = [];

        while (!this.checkToken(TokenType.RPARAM)) {
            do {
                let argName: string = "";
                let argType: Type;
                let initValue: Expression | null = null;
                argName = this.consume(
                    TokenType.IDENTIFIER,
                    "Expected argument name"
                ).value;
                this.consume(TokenType.COLON);
                argType = new Type(
                    this.consume(
                        TokenType.IDENTIFIER,
                        "Expected argument type"
                    ).value
                );
                if (this.matchTokens([TokenType.EQUAL])) {
                    initValue = this.primary();
                }
                args.push(
                    new FunctionDeclarationExpressionArgument(
                        argName,
                        initValue,
                        argType
                    )
                );
            } while (this.matchTokens([TokenType.COMMA]));
        }

        this.consume(TokenType.RPARAM);
        if (this.matchTokens([TokenType.COLON])) {
            let returnType = this.consume(
                TokenType.IDENTIFIER,
                "Expected function return type"
            ).value;
            //console.log(returnType);
            type = new Type(`function<${returnType}>`);
        }
        block = this.statement();
        if (name === null) {
            let anonFunctionName: string = this.generateAnonymousFunction(
                startToken,
                type,
                args,
                block
            );
            let variableExpression = new VariableExpression(startToken, {
                value: anonFunctionName,
                type: TokenType.IDENTIFIER,
            } as Token);
            variableExpression.expectSemiColon = false;
            return variableExpression;
        } else {
            return new FunctionDeclarationExpression(
                startToken,
                name,
                args,
                type,
                block
            );
        }
    }
    synchronize() {
        this.advanceTokenIndex();
        while (!this.isAtEnd()) {
            if (this.previousToken().type === TokenType.SEMICOLON) {
                return;
            }
            switch (this.currentToken().type) {
                case TokenType.FUNCTION:
                case TokenType.VARIABLE:
                case TokenType.CONST:
                case TokenType.RETURN:
                case TokenType.CLASS:
                case TokenType.FOR:
                case TokenType.WHILE:
                case TokenType.IF:
                case TokenType.ELSE:
                    return;
            }
            this.advanceTokenIndex();
        }
    }

    // All helpful functions
    consume(expected: TokenType, message: string = "") {
        let nmessage = "";
        if (!message) {
            nmessage = `Expected '${TokenType[expected]}' but got '${
                TokenType[this.currentToken().type]
            }'`;
        }
        if (this.isAtEnd()) {
            ThrowFatalError(
                message,
                this.previousToken().line,
                this.previousToken().column
            );
        }
        if (this.currentToken().type == expected) {
            this.advanceTokenIndex();

            return this.previousToken();
        } else {
            ThrowFatalError(
                message,
                this.previousToken().line,
                this.previousToken().column
            );
            return this.previousToken(); // will never get here, but typescript won't shut up
        }
    }
    matchTokens(TokenTypes: Array<TokenType>): boolean {
        for (let i = 0; i < TokenTypes.length; i++) {
            if (this.checkToken(TokenTypes[i])) {
                this.advanceTokenIndex();
                return true;
            }
        }
        return false;
    }
    checkToken(expected: TokenType) {
        if (this.isAtEnd()) {
            return false;
        }
        return this.currentToken().type === expected;
    }
    advanceTokenIndex(): Token {
        if (!this.isAtEnd()) this.currentTokenIndex++;
        return this.previousToken();
    }
    currentToken(): Token {
        return this.tokens[this.currentTokenIndex];
    }
    previousToken(): Token {
        return this.tokens[this.currentTokenIndex - 1];
    }
    nextToken(): Token {
        return this.tokens[this.currentTokenIndex + 1];
    }
    isAtEnd(): boolean {
        return this.currentTokenIndex >= this.tokens.length;
    }
}
